src/zope/interface/_zope_interface_coptimizations.c

```diff
@@ -91,6 +95,41 @@ import_declarations(void)
   return 0;
 }
 
+static PyObject*
+_next_super_class(PyObject* ob)
+{
+    // Given an *ob* of type ``super``, return the next
+    // class in the MRO to examine, starting from where the ``super``
+    // was invoked.
+    // Returns a BORROWED reference.
+
+    PyObject* result;
+    PyObject* class_that_invoked_super;
+    PyObject* mro;
+    Py_ssize_t index;
+
+    result = class_that_invoked_super = mro = NULL;
+    #define _NSCHECK(ob, bad_val) do { if (ob == bad_val) { goto EXIT; } } while (0)
+
+    class_that_invoked_super = PyObject_GetAttr(ob, str__thisclass__);
+    _NSCHECK(class_that_invoked_super, NULL);
+
+    mro = PyObject_CallMethodObjArgs(class_that_invoked_super, strmro, NULL);
+    _NSCHECK(mro, NULL);
+
+    index = PySequence_Index(mro, class_that_invoked_super);
```

Comment:
I think that the `mro` must be determined from `ob.__self_class__` not from `ob.__thisclass__`. I expect that `index` above will always be `0`.

Reply:
I'm afraid you have that backwards. In `super(Class, self)` `__self_class__` is always the same as `type(self)`. It never changes and is the most derived class in the MRO (index 0).  It's `__thisclass__` that changes to reflect where in the MRO the new starting search point actually is.

```pycon
>>> class A(object):
...     pass
...
>>> class B(A):
...     pass
...
>>> class C(B):
...     pass
...
>>> class D(C):
...     pass
...
>>> D.__mro__
(<class '__main__.D'>,
 <class '__main__.C'>,
 <class '__main__.B'>,
 <class '__main__.A'>,
 <class 'object'>)
>>> d = D()
>>> sd = super(D, d)
>>> sc = super(C, d)
>>> sb = super(B, d)

>>> (D, sd.__self_class__, sc.__self_class__, sb.__self_class__)
(<class '__main__.D'>,
 <class '__main__.D'>,
 <class '__main__.D'>,
 <class '__main__.D'>)

>>> sd.__thisclass__
<class '__main__.D'>
>>> sc.__thisclass__
<class '__main__.C'>
>>> sb.__thisclass__
<class '__main__.B'>
```

Reply:
>> I'm afraid you have that backwards. In `super(Class, self)` `__self_class__` is always the same as `type(self)`. It never changes and is the most derived class in the MRO (index 0). It's `__thisclass__` that changes to reflect where in the MRO the new starting search point actually is.

> I am aware of this. That's why I am (quite) sure that
> what you need is
```
> mro = PyObject_CallMethodObjArgs(__self_class__, strmro, NULL);
> _NSCHECK(mro, NULL);
>
> index = PySequence_Index(mro, class_that_invoked_super);
```

> With your code, `index` will always be 0 as for any class
> it is the 0's element of its "mro".

Essentially true. We're letting `super` do the harder work of figuring out exactly where in MRO we are. I thought that the class linearization order, which preserves local precedence order, would handle everything else correctly while minimizing our search time here.  But I forgot about diamonds!

Continuing the example from above it looks good:
```pycon
>>> from zope.interface.declarations import _next_super_class
>>> _next_super_class(sd) # super(D, d) means begin looking above D
<class '__main__.C'>
>>> _next_super_class(sc)
<class '__main__.B'>
>>> _next_super_class(sb)
<class '__main__.A'>
```

Introducing a diamond shows the inconsistency:
```pycon
>>> class BB(A):
...     pass
...
>>> class E(B, BB):
...     pass
...
>>> E.__mro__
(<class '__main__.E'>,
 <class '__main__.B'>,
 <class '__main__.BB'>,
 <class '__main__.A'>,
 <class 'object'>)
>>> e = E()
>>> se = super(E, e)
>>> sb = super(B, e)
>>> sbb = super(BB, e)
>>> _next_super_class(se)
<class '__main__.B'>
>>> _next_super_class(sb) # Should be BB
<class '__main__.A'>
>>> _next_super_class(sbb)
<class '__main__.A'>
```

Redefining as suggested is consistent. Thank you!
```pycon
>>> def _next_super_class(ob):
...     # When ``ob`` is an instance of ``super``, return
...     # the next class in the MRO that we should actually be
...     # looking at.
...     class_that_invoked_super = ob.__self_class__
...     mro = class_that_invoked_super.__mro__
...     next_class = mro[mro.index(ob.__thisclass__) + 1]
...     return next_class
…
>>> _next_super_class(se)
<class '__main__.B'>
>>> _next_super_class(sb)
<class '__main__.BB'>
>>> _next_super_class(sbb)
<class '__main__.A'>
```